proposals/content-layer.md

```diff
@@ -0,0 +1,345 @@
+
+Content layer is designed to be a successor to content collections that addresses these limitations and opens up more use cases. It is inspired by the Gatsby data layer, but with a simpler API and no GraphQL or graph data store.
+
+# Goals
+
+- Create a successor to content collections that can be used with local and remote data.
+- Decouple content from Vite, so that data is no longer implemented as Rollup chunks.
+- Provide a simple API for defining collections with a migration path from content collections.
+- Support local files in user-defined locations with built-in file and glob loaders.
+- Initially support Markdown rendering and JSON data for local files, with support for other formats in the future.
+- Provide an API for defining custom loaders that is powerful enough to handle a wide range of use cases, including CMS integrations.
+- Provide a higher-level API for simple inline loaders.
+- Make the API scalable to tens of thousands of entries, with good performance and low memory usage.
+- Allow loaders to define their own schemas, including dynamic schemas that can be introspected from the data source.
+
+# Non-Goals
+
+- Non-goal: allowing loaders to define multiple collections automatically. e.g. separate collections would need to be manually defined for posts and categories in a blog.
+- Non-goal: dependency tracing for entries.
+- Out of scope: hot-reloading remote data.
+- Out of scope: rendering markdown from remote data. A loader could store rendered HTML, but it would be up to the loader to handle this.
+- Out of scope: custom `Content` components.
+- Future: support for Markdoc and MDX rendering.
+- Future: SQLite-based backend for collections.
+- Future: support for queries more complex than get by ID.
+
+# Detailed Design
+
+## Collection Definition
+
+Collections are defined in a similar way to current content collections, using `defineCollection()` in `src/content/config.ts`. There are two new collection types: `experimental_content` and `experimental_data`. The only difference between them is that when a collection is defined as `experimental_content`, entries will have a `render()` method that generates a component for rendering HTML content. This requires the loader to have set the `rendered.html` property.
+
+The `reference()` helper can be used in the same way as content collections, to reference other collections.
+
+## Built-in loaders
+
+There are two built-in loaders: `file()` and `glob()`, which load data from the local filesystem. The `glob()` loader covers the current use case of directories full of markdown or JSON content. The `glob()` helper is more flexible than in current content collections, as it can load data from anywhere on the filesystem. The `file()` loader loads multiple entries from a single file. Both loaders can process markdown in the same way as content collections. They can also extract images in the same way as content collections.
```

Comment:
My suggestion: I would change the narrative of the RFC like this:
1. Start by showing how to define content and render it, in the *most minimal* way. This is an excuse to show how **simple** is the new API, because a user can use **a simple JS function** to define some data.  
2. From there, let's talk about the `loader` API and explain it. Talk about its design (astro/vite-like), then let's get technical and explain the other APIs, e.g. `watcher`, `store`, etc. We will talk about the fact that the `store` is scoped to the current collection. Here we can be as technical as we want. We can also talk about "behind the scenes" business logic, if it is relevant to us developers, to provide context when reviewing the PRs.
3. Eventually, we can **mention** our built-in loaders, but I believe there's no need to go into detail too much since they use the APIs that were explained before. However, I believe it's worth mentioning how `glob` **could be used** for an almost built-in replacement of the current content collections.